### PR TITLE
closes #2, propagate typechecking through vectorizing transformations

### LINKED ISSUE #2: propagate `pytree_dataclass` typechecking through vectorizing transformations
A rough sketch of the way a transformer is implemented in `train.py` might be 
```python
@pytree_dataclass
class Model:
  w_up: f32['layers d_model/d d_ff/t']
  # ...
  
@typechecked
def transformer_block(x: f32[b'batch/d seq d_model/t'], w_block: Any) -> f32[b'batch/d seq d_model/t']:
  # ...
  return x, ()
  
@typed_shard_map
def transformer(x: f32[b'batch/d seq d_model/t'], w_model: Model) -> f32[b'batch/d seq d_model/t']:
  y, _ = jax.lax.scan(transformer_block, x, w)
  return y
```
It would be helpful if there was some way of checking that `w_block` had the expected type, i.e the same as `Model` but with the leading `'layers'` axis removed from each field. One very hacky way of doing this is to use
```python
@pytree_dataclass
class Block:
  w_up: f32['d_model/d d_ff/t']
  # ...
  
@pytree_dataclass
class Model(Block):
  w_up: f32['layers d_model/d d_ff/t']
  # ...
  
@typechecked
def transformer_block(x: f32[b'batch/d seq d_model/t'], w_block: Block) -> f32[b'batch/d seq d_model/t']:
# ...
```
However, in more complex situations, this quickly leads to a large amount of boilerplate dataclasses. Ideally, there might be some kind of api that would allow e.g
```python
Model = expand_named_axes('layers', Block)
```
If the maintainers think this is a good idea, I'm happy to start working on a draft implementation.

### PR DESCRIPTION
**Pull request by zhangir-azerbayev**
**Motivation**
issue #2 

**Demo**
The code snippet
```python
from dataclasses import fields, is_dataclass

from shardlib.shardtypes import f32, pytree_dataclass, Array
from shardlib import shardtypes
shardtypes.register_with_typeguard()

@pytree_dataclass
class MLP:
  w_up: f32['d_model d_ffn']
  ...

@pytree_dataclass
class TransformerLayer:
  mlp: MLP
  ...

def print_dataclass_fields(cls, indent=''):
  for field in fields(cls):
    field_type = field.type
    print(f'{indent}{field.name}: {field_type}')
    if is_dataclass(field_type):
      print_dataclass_fields(field_type, indent + '  ')


Transformer = Array['layers', TransformerLayer]
print_dataclass_fields(Transformer)
```
has the following output:
```
mlp: <class 'shardlib.shardtypes.MLP'>
  w_up: shardlib.shardtypes.f32[b'layers d_model d_ffn']
```
See changes to `train.Model` for an example of how the api is used in practice.

**Open questions**
- Right now, the output of `shardtypes.Array` is a subclass of the input class and has the same `__name__` attribute. Does this ever cause problems?
- The type returned by `shardtypes.Array` only retains the field names and field types of the input class. Is there other information we want to copy?

### CODE REVIEW

**Comment by tensorpro** [comment]
This looks great! 

To respond to the questions: I think your returning the fields we need, and I don't think re-using the class `__name__` should cause problems. I'll look into this a bit more and do a final review by tommorow.

Thanks!

**Review by tensorpro — APPROVED** [review]

**Review by reinerp — APPROVED** [review]
Thanks for the great contribution! Any other ideas are always welcome.

**Comment by zhangir-azerbayev** [comment]
thanks!
